### PR TITLE
[FIX] web: add the same image on different records

### PR DESCRIPTION
**Pull request by MerlinGuillaume**
Using the same image on new records does not work without refreshing the page

Steps to reproduce:
1. Install Sales
2. Go to Sales > Products > Products and create a new product
3. Set a name and an image on the product
4. Create a new product (with the 'New' button in the top right of the form)
5. Set the image used previously on this product
6. The image does not save

Solution:
Add a `t-key` on the FileUploader in order to recreate a FileUploader when we open a new form

Problem:
The FileUploader component is not reset when we open a new form (the files previously uploaded are still present on the FileUploader). `onFileChange` is thus not called when we use the same image, as the file hasn't changed.

opw-3151756

### CODE REVIEW

**Review by hubvd — APPROVED** [review]
Well done :)

**Comment by MerlinGuillaume** [comment]
Hello @odoo/rd-framework-js, could you review this please ?

File: addons/web/static/src/views/fields/file_handler.js [hunk 1]
```diff
@@ -18,6 +18,10 @@ export class FileUploader extends Component {
         this.state = useState({
             isUploading: false,
         });
+
+        if (this.props.bus) {
+            this.props.bus.on("clear_file_input", this, () => this.fileInputRef.el.value = "");
+        }
```
Comment (brboi) [inline addons/web/static/src/views/fields/file_handler.js:24]: seems weird to share a bus through props for this single use case

File: addons/web/static/src/views/fields/file_handler.js
Comment (brboi) [inline addons/web/static/src/views/fields/file_handler.js:1]: this.id is dead code, could you remove it as well as the nextId static member ?
Thanks

**Review by brboi — CHANGES_REQUESTED** [review]
Your solution is OK, but what do you think about the following ?
```diff
diff --git a/addons/web/static/src/views/fields/file_handler.js b/addons/web/static/src/views/fields/file_handler.js
index b1dc6d38e1e2..4f8b2c996e7b 100644
--- a/addons/web/static/src/views/fields/file_handler.js
+++ b/addons/web/static/src/views/fields/file_handler.js
@@ -6,17 +6,23 @@ import { getDataURLFromFile } from "@web/core/utils/urls";
 import { session } from "@web/session";
 import { formatFloat } from "./formatters";
 
-import { Component, useRef, useState } from "@odoo/owl";
+import { Component, onWillUpdateProps, useRef, useState } from "@odoo/owl";
 
 const DEFAULT_MAX_FILE_SIZE = 128 * 1024 * 1024;
 
 export class FileUploader extends Component {
     setup() {
         this.notification = useService("notification");
-        this.id = `o_fileupload_${++FileUploader.nextId}`;
         this.fileInputRef = useRef("fileInput");
         this.state = useState({
             isUploading: false,
+            key: this.props.key,
+        });
+        onWillUpdateProps((nextProps) => {
+            if (this.state.key !== nextProps.key) {
+                this.fileInputRef.el.value = "";
+                this.state.key = nextProps.key;
+            }
         });
     }
 
@@ -74,4 +80,3 @@ export class FileUploader extends Component {
 }
 
 FileUploader.template = "web.FileUploader";
-FileUploader.nextId = 0;
diff --git a/addons/web/static/src/views/fields/image/image_field.xml b/addons/web/static/src/views/fields/image/image_field.xml
index bb9d42f71875..5ca160a7b29a 100644
--- a/addons/web/static/src/views/fields/image/image_field.xml
+++ b/addons/web/static/src/views/fields/image/image_field.xml
@@ -7,6 +7,7 @@
                 <t t-if="!props.readonly">
                     <FileUploader
                         acceptedFileExtensions="props.acceptedFileExtensions"
+                        key="rawCacheKey"
                         onUploaded.bind="onFileUploaded"
                         type="'image'"
                     >
```


Also, I am requesting changes as a fix should generally come with a test, in order to avoid breaking this again in the future.

**Comment by sdegueldre** [comment]
Hm, I feel like this is a more general problem. Somewhere, somehow, if the record changes the entire view should be created from scratch, we probably want to reset the internal state of all HTML elements in those scenarios. Probably needs a well placed t-key cc @aab-odoo

**Comment by hubvd** [comment]
> Hm, I feel like this is a more general problem

This one could also be fixed this way I guess https://github.com/odoo/odoo/pull/116179

**Comment by aab-odoo** [comment]
> Hm, I feel like this is a more general problem. Somewhere, somehow, if the record changes the entire view should be created from scratch, we probably want to reset the internal state of all HTML elements in those scenarios. Probably needs a well placed t-key cc @aab-odoo

Not sure. In this case we have an issue because the FileUploader doesn't set the value of the input right? It's a very specific and unique case, so maybe @brboi solution is fine.

File: addons/web/static/src/views/fields/file_handler.js [hunk 2]
```diff
@@ -6,17 +6,23 @@ import { getDataURLFromFile } from "@web/core/utils/urls";
...
 export class FileUploader extends Component {
     setup() {
         this.notification = useService("notification");
-        this.id = `o_fileupload_${++FileUploader.nextId}`;
         this.fileInputRef = useRef("fileInput");
         this.state = useState({
             isUploading: false,
+            key: this.props.key,
+        });
+        onWillUpdateProps((nextProps) => {
```
Comment (ged-odoo) [inline addons/web/static/src/views/fields/file_handler.js:21]: you can remove this and use `t-key` in the image field template

File: addons/web/static/src/views/fields/image/image_field.xml [hunk 3]
```diff
@@ -7,6 +7,7 @@
                 <t t-if="!props.readonly">
                     <FileUploader
                         acceptedFileExtensions="props.acceptedFileExtensions"
+                        key="props.record.resId"
```
Comment (ged-odoo) [inline addons/web/static/src/views/fields/image/image_field.xml:10]: and here, you can use `t-key`:

```
                        t-key="props.record.resId"

```

**Review by brboi — APPROVED** [review]
robodoo r+

**Pull request closed without merging**